### PR TITLE
WIP: 2020 12 09 box keypair

### PR DESCRIPTION
**Pull request by thedavidmeister**

### CODE REVIEW

File: crates/lair_keystore_api/src/internal/crypto_box.rs [hunk 1]
```diff
@@ -0,0 +1,179 @@
+use std::sync::Arc;
```
Comment (thedavidmeister) [inline crates/lair_keystore_api/src/internal/crypto_box.rs:1]: this file has the juicy bits in it

File: crates/lair_keystore_api/src/internal/crypto_box.rs [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+        Self {
+            data: Arc::new(v)
+        }
+    }
+}
+
+/// Wrapper around crypto_box from whatever lib we use.
+/// No BYO nonces. Nonces always random and returned as part of `CryptoBoxEncryptedData`.
+/// No BYO algorithms (cipher agility). Algorithm always X25519XSalsa20Poly1305.
+/// Currently no additional associated data but DNA space may be included in the future.
```
Comment (neonphog) [inline crates/lair_keystore_api/src/internal/crypto_box.rs:120]: This makes it sound like we will never allow flexibility...

I'm totally cool with our first cut providing a minimal api, but eventually I'd like to just expose as much of the libsodium api as we can, and let devs decide how complex they want to make things... we can totally make options with a ::default() that fills in a None for nonce that causes a random nonce to be generated, but let people fill that out if the care for some reason.

Similarly - we don't know that dna is the only associated data people will ever want..

&& eventually, the only api I really want is `crypto_aead_xchacha20poly1305_ietf_*`

To be clear, I'm only asking for a documentation change right now to not make it seem so final.

File: crates/lair_keystore_api/src/internal/crypto_box.rs [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+/// @see https://eprint.iacr.org/2019/519.pdf for 'context separable interfaces'
+pub fn crypto_box(sender: x25519::X25519PrivKey, recipient: x25519::X25519PubKey, data: Arc<CryptoBoxData>) -> crate::error::LairResult<CryptoBoxEncryptedData> {
+    use lib_crypto_box::aead::Aead;
+    let sender_box = lib_crypto_box::SalsaBox::new(recipient.as_ref(), sender.as_ref());
+    let nonce = CryptoBoxNonce::new_random();
+
+    let mut padded_data = data.data.to_vec();
+    block_padding::Iso7816::pad_block(&mut padded_data, BLOCK_PADDING_SIZE)?;
+
+    let encrypted_data = Arc::new(sender_box.encrypt(AsRef::<[u8; NONCE_BYTES]>::as_ref(&nonce).into(), padded_data.as_slice())?);
```
Comment (neonphog) [inline crates/lair_keystore_api/src/internal/crypto_box.rs:156]: I think this whole function (but especially this line) should be run in a rayon scope.

File: crates/lair_keystore_api/src/internal/crypto_box.rs [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    })
+}
+
+/// Wrapper around crypto_box_open from whatever lib we use.
+/// Exact inverse of `crypto_box_open` so nonce must be provided in `CryptoBoxEncryptedData`.
+/// The recipient's private key encrypts _from_ the sender's pubkey.
+pub fn crypto_box_open(recipient: x25519::X25519PrivKey, sender: x25519::X25519PubKey, encrypted_data: Arc<CryptoBoxEncryptedData>) -> crate::error::LairResult<CryptoBoxData> {
+    use lib_crypto_box::aead::Aead;
+    let recipient_box = lib_crypto_box::SalsaBox::new(sender.as_ref(), recipient.as_ref());
+    let decrypted_data = recipient_box.decrypt(AsRef::<[u8; NONCE_BYTES]>::as_ref(&encrypted_data.nonce).into(), encrypted_data.encrypted_data.as_slice())?;
```
Comment (neonphog) [inline crates/lair_keystore_api/src/internal/crypto_box.rs:172]: run this function in a rayon scope

File: crates/lair_keystore_api/src/internal/wire.rs [hunk 5]
```diff
@@ -494,6 +502,303 @@ macro_rules! wire_type_meta_macro {
...
+                let mut writer = default_encode_setup!(msg_id, wire_type);
+                writer.write_u32(**keystore_index)?;
+                writer.write_bytes_exact(AsRef::<[u8]>::as_ref(pub_key), 32)?;
+                Ok(writer.into_vec())
+            } |reader| {
+                let msg_id = reader.read_u64()?;
+                let keystore_index = reader.read_u32()?;
+                let pub_key = reader.read_bytes(32)?;
+                let mut pub_key_array = [0; 32];
+                pub_key_array.copy_from_slice(&pub_key);
```
Comment (neonphog) [inline crates/lair_keystore_api/src/internal/wire.rs:526]: Not a big deal, but if you wanted to, there's already an extension trait on this reader, you could add a fn that reads directly into an existing buffer and skip this additional copy.

File: crates/lair_keystore_api/src/internal/x25519.rs [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+        self.to_bytes().hash(state)
+    }
+}
+
+/// Generate a new random x25519 keypair.
+pub async fn x25519_keypair_new_from_entropy() -> LairResult<entry::EntryX25519> {
+    rayon_exec(move || {
+        // This is fine _for use with the `crypto_box` crate_ because they specify the `CryptoRng`
+        // trait on the `generate()` method below.
+        // @todo store the rng somewhere for efficiency.
```
Comment (neonphog) [inline crates/lair_keystore_api/src/internal/x25519.rs:161]: I think TLS is actually pretty efficient... not sure we really need anything more than that.

**Comment by neonphog** [comment]
looking good @thedavidmeister :+1: - lemme know when you're ready and I can give it a final look-through.

File: crates/lair_keystore_api/src/internal/crypto_box.rs (hunk 2 above)
Comment (thedavidmeister) [reply crates/lair_keystore_api/src/internal/crypto_box.rs:120]: @neonphog i agree 100% not my intent to make this sound set in stone

File: crates/lair_keystore_api/src/internal/crypto_box.rs (hunk 3 above)
Comment (thedavidmeister) [reply crates/lair_keystore_api/src/internal/crypto_box.rs:156]: @neonphog nice

File: crates/lair_keystore_api/src/internal/wire.rs (hunk 5 above)
Comment (thedavidmeister) [reply crates/lair_keystore_api/src/internal/wire.rs:526]: oh yeah, also i implemented `TryFrom` already and forgot about it here :sweat_smile:

**Review by neonphog — APPROVED** [review]
suggest potentially adding some negative tests showing that open errors with corrupt data.

But, LGTM :+1:

**Pull request merged**